core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -853,35 +853,50 @@ public boolean deleteRetentionTimeBreachedSegments(RemoteLogSegmentMetadata meta
                 return isSegmentDeleted;
             }
 
-            private boolean deleteLogStartOffsetBreachedSegments(RemoteLogSegmentMetadata metadata, long startOffset)
+            private boolean deleteLogStartOffsetBreachedSegments(RemoteLogSegmentMetadata metadata,
+                                                                 long logStartOffset,
+                                                                 NavigableMap<Integer, Long> leaderEpochEntries)
                     throws RemoteStorageException, ExecutionException, InterruptedException {
-                boolean isSegmentDeleted = deleteRemoteLogSegment(metadata, x -> startOffset > x.endOffset());
-                if (isSegmentDeleted && retentionSizeData.isPresent()) {
-                    remainingBreachedSize = Math.max(0, remainingBreachedSize - metadata.segmentSizeInBytes());
-                    logger.info("Deleted remote log segment {} due to log start offset {} breach", metadata.remoteLogSegmentId(), startOffset);
+                boolean isSegmentDeleted = deleteRemoteLogSegment(metadata, ignored -> {
+                    if (!leaderEpochEntries.isEmpty()) {
+                        // Note that `logStartOffset` and `leaderEpochEntries.firstEntry().getValue()` should be same
+                        Integer firstEpoch = leaderEpochEntries.firstKey();
+                        return metadata.segmentLeaderEpochs().keySet().stream().allMatch(epoch -> epoch <= firstEpoch)
+                                && metadata.endOffset() < logStartOffset;
+                    }
+                    return false;
+                });
+                if (isSegmentDeleted) {
+                    remainingBreachedSize = retentionSizeData.isPresent() ?
```

Comment:
It seems we should not change `remainingBreachedSize` as that is computed only with the segments that are in the current leader epoch lineage.

Reply:
Good point! Addressed your review comment. PTAL.